// @sudoku/stores/grid.js
import { BOX_SIZE, SUDOKU_SIZE } from '@sudoku/constants';
import { decodeSencode, encodeSudoku } from '@sudoku/sencode';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { derived, writable, get } from 'svelte/store';
import { hints } from './hints';

function createGrid() {
    const grid = writable(Array(9).fill(0).map(() => Array(9).fill(0)));
    return {
        subscribe: grid.subscribe,
        generate(difficulty) { grid.set(generateSudoku(difficulty)); },
        decodeSencode(sencode) { grid.set(decodeSencode(sencode)); },
        get(gridStore, x, y) { return gridStore[y][x]; },
        getSencode(gridStore) { return encodeSudoku(gridStore); },
    };
}
export const grid = createGrid();

function createUserGrid() {
    const userGrid = writable(Array(9).fill(0).map(() => Array(9).fill(0)));
    grid.subscribe($grid => {
        let newGrid = $grid.map(row => [...row]);
        userGrid.set(newGrid);
    });

    return {
        subscribe: userGrid.subscribe,
        set: (pos, value) => {
            userGrid.update($userGrid => {
                $userGrid[pos.y][pos.x] = value;
                return [...$userGrid];
            });
        },
        applyHint: (pos) => {
            hints.useHint();
            userGrid.update($userGrid => {
                const solvedSudoku = solveSudoku($userGrid);
                $userGrid[pos.y][pos.x] = solvedSudoku[pos.y][pos.x];
                return [...$userGrid];
            });
        },
        getSencode: () => encodeSudoku(get(userGrid)),
        decodeSencode: (sencode) => userGrid.set(decodeSencode(sencode)),
        getCandidates: (pos) => {
            const currentGrid = get(userGrid);
            const candidates = [];
            for (let num = 1; num <= 9; num++) {
                if (isMoveValid(currentGrid, pos.x, pos.y, num)) candidates.push(num);
            }
            return candidates;
        }
    };
}

function isMoveValid(grid, x, y, num) {
    for (let i = 0; i < SUDOKU_SIZE; i++) {
        if (grid[y][i] === num || grid[i][x] === num) return false;
    }
    const startY = Math.floor(y / BOX_SIZE) * BOX_SIZE;
    const startX = Math.floor(x / BOX_SIZE) * BOX_SIZE;
    for (let row = startY; row < startY + BOX_SIZE; row++) {
        for (let col = startX; col < startX + BOX_SIZE; col++) {
            if (grid[row][col] === num) return false;
        }
    }
    return true;
}

export const userGrid = createUserGrid();

export const invalidCells = derived(userGrid, $userGrid => {
    const _invalidCells = [];
    const addInvalid = (x, y) => { const xy = x + ',' + y; if (!_invalidCells.includes(xy)) _invalidCells.push(xy); };
    for (let y = 0; y < SUDOKU_SIZE; y++) {
        for (let x = 0; x < SUDOKU_SIZE; x++) {
            const value = $userGrid[y][x];
            if (value) {
                for (let i = 0; i < SUDOKU_SIZE; i++) {
                    if (i !== x && $userGrid[y][i] === value) addInvalid(x, y);
                    if (i !== y && $userGrid[i][x] === value) addInvalid(x, y);
                }
                const startY = Math.floor(y / BOX_SIZE) * BOX_SIZE;
                const startX = Math.floor(x / BOX_SIZE) * BOX_SIZE;
                for (let row = startY; row < startY + BOX_SIZE; row++) {
                    for (let col = startX; col < startX + BOX_SIZE; col++) {
                        if (row !== y && col !== x && $userGrid[row][col] === value) addInvalid(col, row);
                    }
                }
            }
        }
    }
    return _invalidCells;
}, []);
