// @sudoku/stores/grid.js
import { BOX_SIZE, SUDOKU_SIZE } from '@sudoku/constants';
import { decodeSencode, encodeSudoku } from '@sudoku/sencode';
import { generateSudoku, solveSudoku } from '@sudoku/sudoku';
import { derived, writable, get } from 'svelte/store';
import { hints } from './hints';

// --- 原有逻辑 ---
function createGrid() {
    const grid = writable(Array(9).fill(0).map(() => Array(9).fill(0)));
    return {
        subscribe: grid.subscribe,
        generate(difficulty) { grid.set(generateSudoku(difficulty)); },
        decodeSencode(sencode) { grid.set(decodeSencode(sencode)); },
        get: (gridStore, x, y) => gridStore[y][x],
        getSencode: (gridStore) => encodeSudoku(gridStore),
    };
}
export const grid = createGrid();

function createUserGrid() {
    const userGrid = writable(Array(9).fill(0).map(() => Array(9).fill(0)));
    grid.subscribe($grid => {
        userGrid.set($grid.map(row => [...row]));
    });

    return {
        subscribe: userGrid.subscribe,
        set: (pos, value) => {
            userGrid.update($userGrid => {
                $userGrid[pos.y][pos.x] = value;
                return [...$userGrid.map(row => [...row])];
            });
        },
        // ... (其他原有方法如 applyHint, getCandidates 等保持不变)
    };
}
export const userGrid = createUserGrid();

// --- 补全的测试契约接口 ---

export function createSudoku(initialGrid = null) {
    // 防御性复制：确保返回的是一个独立的 grid 副本
    let currentGrid = initialGrid 
        ? initialGrid.map(row => [...row]) 
        : Array(9).fill(0).map(() => Array(9).fill(0));

    return {
        getGrid: () => currentGrid.map(row => [...row]),
        guess: (row, col, value) => { currentGrid[row][col] = value; },
        toString: () => currentGrid.map(row => row.join('')).join('\n'),
        toJSON: () => currentGrid.map(row => [...row]),
        clone: () => createSudoku(currentGrid)
    };
}

export function createGame() {
    let history = []; // 存储状态栈
    let redoStack = [];
    let state = createSudoku();

    return {
        // 暴露 createSudoku 的方法
        ...state,
        undo: () => {
            if (history.length > 0) {
                redoStack.push(state.toJSON());
                state = createSudoku(history.pop());
            }
        },
        redo: () => {
            if (redoStack.length > 0) {
                history.push(state.toJSON());
                state = createSudoku(redoStack.pop());
            }
        },
        guess: (row, col, value) => {
            history.push(state.toJSON());
            redoStack = []; // 新操作清空重做栈
            state.guess(row, col, value);
        }
    };
}
